File: pdf_parser.py
import pdfplumber
import re
import math

def clean_pdf_text(text):
    if not text:
        return ""
    text = text.replace('\x00', '').replace('\r', '')
    # Normalize unicode dash variations (en-dash, em-dash, figure dash, etc.) to standard ASCII hyphen '-'
    text = re.sub(r'[\u2010\u2011\u2012\u2013\u2014\u2015\u2212]', '-', text)
    return text

def parse_course_code_name(asig_str):
    asig_str = clean_pdf_text(asig_str.strip())
    asig_str = re.sub(r'^[-\s]+', '', asig_str)
    
    # Try splitting by ' - ' or '-' with optional whitespace
    parts = re.split(r'\s*-\s*', asig_str, maxsplit=1)
    if len(parts) == 2 and re.match(r'^[A-Za-z0-9]{3,12}$', parts[0].strip()):
        return parts[0].strip(), parts[1].strip()
    
    m = re.match(r'^([A-Za-z0-9]{3,12})\s*-\s*(.*)$', asig_str)
    if m:
        return m.group(1).strip(), m.group(2).strip()
    
    m2 = re.match(r'^([A-Za-z0-9]{5,12})\s+(.*)$', asig_str)
    if m2:
        return m2.group(1).strip(), m2.group(2).strip()

    return "", asig_str

def extract_header_metadata(text_lines):
    meta = {
        'student_code': '',
        'student_name': '',
        'facultad': '',
        'escuela': '',
        'plan': ''
    }
    for line in text_lines:
        line_s = line.strip()
        if ('Código de Matrícula' in line_s or 'Cdigo de Matrcula' in line_s) and ':' in line_s:
            parts = line_s.split(':', 1)
            meta['student_code'] = parts[1].strip()
        elif 'Nombres y Apellidos' in line_s and ':' in line_s:
            parts = line_s.split(':', 1)
            meta['student_name'] = parts[1].strip()
        elif 'Facultad' in line_s and ':' in line_s:
            parts = line_s.split(':', 1)
            meta['facultad'] = parts[1].strip()
        elif 'Escuela' in line_s and ':' in line_s:
            parts = line_s.split(':', 1)
            meta['escuela'] = parts[1].strip()
        elif 'Plan' in line_s and ':' in line_s and ('Plan De' in line_s or '202' in line_s):
            parts = line_s.split(':', 1)
            meta['plan'] = parts[1].strip()
    return meta

def parse_historial_file(pdf_file_or_path):
    """
    Parses an UNMSM Historial Académico PDF file.
    Returns (periods, resumen, metadata).
    """
    periods = []
    current_period = None
    resumen = {
        'required_credits': 221.0,
        'approved_credits': 0.0,
        'obligatorios': 0.0,
        'especialidad': 0.0,
        'electivos_generales': 0.0,
        'electivos_especialidad': 0.0,
        'optativos': 0.0,
        'alternativos': 0.0,
        'otra_especialidad': 0.0,
        'mas_de_una_vez': 0.0,
        'otros': 0.0,
        'missing_credits': 221.0,
        'ppg': 0.0
    }
    metadata = {}

    with pdfplumber.open(pdf_file_or_path) as pdf:
        for page_idx, page in enumerate(pdf.pages):
            raw_text = page.extract_text() or ""
            text = clean_pdf_text(raw_text)
            lines = text.split('\n')
            
            if page_idx == 0:
                metadata = extract_header_metadata(lines)
            
            for l in lines:
                l_str = l.strip()
                if 'Creditaje Requerido para Egresar' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['required_credits'] = float(nums[-1])
                elif 'Creditaje Apobrado' in l_str or 'Creditaje Aprobado' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['approved_credits'] = float(nums[-1])
                elif 'Obligatorios' in l_str and 'Creditaje' not in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['obligatorios'] = float(nums[-1])
                elif 'De Especialidad' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['especialidad'] = float(nums[-1])
                elif 'Electivos Generales' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['electivos_generales'] = float(nums[-1])
                elif 'Electivos de Especialidad' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['electivos_especialidad'] = float(nums[-1])
                elif 'Optativos' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['optativos'] = float(nums[-1])
                elif 'Alternativos' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['alternativos'] = float(nums[-1])
                elif 'De Otra Especialidad' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['otra_especialidad'] = float(nums[-1])
                elif 'Más de una vez' in l_str or 'Ms de una vez' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['mas_de_una_vez'] = float(nums[-1])
def parse_historial_file(pdf_file_or_path):
    """
    Parses an UNMSM Historial Académico PDF file.
    Returns (periods, resumen, metadata).
    """
    periods = []
    current_period = None
    resumen = {
        'required_credits': 221.0,
        'approved_credits': 0.0,
        'obligatorios': 0.0,
        'especialidad': 0.0,
        'electivos_generales': 0.0,
        'electivos_especialidad': 0.0,
        'optativos': 0.0,
        'alternativos': 0.0,
        'otra_especialidad': 0.0,
        'mas_de_una_vez': 0.0,
        'otros': 0.0,
        'missing_credits': 221.0,
        'ppg': 0.0
    }
    metadata = {}

    with pdfplumber.open(pdf_file_or_path) as pdf:
        for page_idx, page in enumerate(pdf.pages):
            raw_text = page.extract_text() or ""
            text = clean_pdf_text(raw_text)
            lines = text.split('\n')
            
            if page_idx == 0:
                metadata = extract_header_metadata(lines)
            
            for l in lines:
                l_str = l.strip()
                if 'Creditaje Requerido para Egresar' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['required_credits'] = float(nums[-1])
                elif 'Creditaje Apobrado' in l_str or 'Creditaje Aprobado' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['approved_credits'] = float(nums[-1])
                elif 'Obligatorios' in l_str and 'Creditaje' not in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['obligatorios'] = float(nums[-1])
                elif 'De Especialidad' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['especialidad'] = float(nums[-1])
                elif 'Electivos Generales' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['electivos_generales'] = float(nums[-1])
                elif 'Electivos de Especialidad' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['electivos_especialidad'] = float(nums[-1])
                elif 'Optativos' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['optativos'] = float(nums[-1])
                elif 'Alternativos' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['alternativos'] = float(nums[-1])
                elif 'De Otra Especialidad' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['otra_especialidad'] = float(nums[-1])
                elif 'Más de una vez' in l_str or 'Ms de una vez' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['mas_de_una_vez'] = float(nums[-1])
                elif 'Otros' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['otros'] = float(nums[-1])
                elif 'Creditaje Faltante' in l_str:
                    nums = re.findall(r'\d+\.?\d*', l_str)
                    if nums: resumen['missing_credits'] = float(nums[-1])
                elif 'Promedio Ponderado' in l_str and 'General' not in l_str:
                    m = re.findall(r'\d+\.\d+', l_str)
                    if m: resumen['ppg'] = float(m[-1])

            pending_course = None

            for line in lines:
                line_s = line.strip()
                m_period = re.search(r'Periodo Acad[eé]mico\s+(\d{4}-\d)', line_s, re.IGNORECASE)
                if m_period:
                    p_name = m_period.group(1)
                    current_period = next((p for p in periods if p['period'] == p_name), None)
                    if not current_period:
                        current_period = {'period': p_name, 'courses': []}
                        periods.append(current_period)
                    pending_course = None
                    continue
                
                if not current_period:
                    continue

                m_start = re.match(r'^(\d+)\s+(\d{4})\s+([OEΟ])(?:\s+(.*))?$', line_s)
                if m_start:
                    ciclo, plan, tipo, rest = m_start.groups()
                    rest = (rest or '').strip()
                    # Strictly match standard UNMSM acta ending [PA]\s*-\s*\S+
                    m_end = re.search(r'(?:(\d{1,2})\s+)?(\d+\.\d)\s+(\d+)\s+([PA]\s*-\s*\S+)$', rest) if rest else None
                    if m_end:
                        calif, cred, sec, acta = m_end.groups()
                        asig = rest[:m_end.start()].strip()
                        calif_val = int(calif) if calif else None
                        code, name = parse_course_code_name(asig)

                        current_period['courses'].append({
                            'id': f"{p_name}_{code}_{sec}",
                            'ciclo': int(ciclo),
                            'plan': plan,
                            'tipo': tipo,
                            'codigo': code,
                            'nombre': name,
                            'asignatura_full': asig,
                            'calificacion': calif_val,
                            'creditos': float(cred),
                            'seccion': sec,
                            'acta': acta,
                            'ep': calif_val if calif_val is not None else 0,
                            'ec': calif_val if calif_val is not None else 0,
                            'ef': calif_val if calif_val is not None else 0,
                            'en_curso': (calif_val is None),
                            'user_edited': False
                        })
                        pending_course = None
                    else:
                        pending_course = {'ciclo': int(ciclo), 'plan': plan, 'tipo': tipo, 'prefix': rest}
                elif pending_course:
                    m_end = re.search(r'(?:(\d{1,2})\s+)?(\d+\.\d)\s+(\d+)\s+([PA]\s*-\s*\S+)$', line_s)
                    if m_end:
                        calif, cred, sec, acta = m_end.groups()
                        suffix = line_s[:m_end.start()].strip()
                        full_asig = (pending_course['prefix'] + " " + suffix).strip()
                        calif_val = int(calif) if calif else None
                        code, name = parse_course_code_name(full_asig)

                        current_period['courses'].append({
                            'id': f"{current_period['period']}_{code}_{sec}",
                            'ciclo': pending_course['ciclo'],
                            'plan': pending_course['plan'],
                            'tipo': pending_course['tipo'],
                            'codigo': code,
                            'nombre': name,
                            'asignatura_full': full_asig,
                            'calificacion': calif_val,
                            'creditos': float(cred),
                            'seccion': sec,
                            'acta': acta,
                            'ep': calif_val if calif_val is not None else 0,
                            'ec': calif_val if calif_val is not None else 0,
                            'ef': calif_val if calif_val is not None else 0,
                            'en_curso': (calif_val is None),
                            'user_edited': False
                        })
                        pending_course = None
                    else:
                        pending_course['prefix'] = (pending_course['prefix'] + " " + line_s).strip()

    return periods, resumen, metadata

    return periods, resumen, metadata

def parse_plan_file(pdf_file_or_path):
    """
    Parses an UNMSM Plan de Estudios PDF file.
    Returns (cycles, metadata).
    """
    cycles = {}
    metadata = {}
    current_detected_cycle = 1
    
    with pdfplumber.open(pdf_file_or_path) as pdf:
        for page_idx, page in enumerate(pdf.pages):
            raw_text = page.extract_text() or ""
            text = clean_pdf_text(raw_text)
            lines = text.split('\n')
            
            if page_idx == 0:
                metadata = extract_header_metadata(lines)

            tables = page.extract_tables()
            
            for table in tables:
                for row in table:
                    if not row or len(row) < 4:
                        continue
                    row_str_full = " ".join([clean_pdf_text(str(cell or '')) for cell in row])
                    if 'Esp.' in row_str_full and 'Asignatura' in row_str_full:
                        continue
                    
                    esp = clean_pdf_text(str(row[0] or '')).strip()
                    asig_str = clean_pdf_text(str(row[1] or '')).replace('\n', ' ').strip()
                    cred_str = clean_pdf_text(str(row[2] or '')).strip()
                    tipo_str = clean_pdf_text(str(row[3] or '')).strip()
                    grupo_str = clean_pdf_text(str(row[4] or '')).strip() if len(row) > 4 else '--'
                    prereq_str = clean_pdf_text(str(row[5] or '')).replace('\n', ' ').strip() if len(row) > 5 else '--'
                    prereq_str = re.sub(r'\s+', ' ', prereq_str)
                    
                    if not asig_str or not cred_str:
                        continue
                    
                    try:
                        cred = float(cred_str)
                    except ValueError:
                        continue
                    
                    code, name = parse_course_code_name(asig_str)

                    c_num = current_detected_cycle
                    # Generic UNMSM 9-digit code regex: \d{5}(\d{2})\d{2} -> 6th and 7th digits are the cycle number!
                    m_code_cycle = re.search(r'\d{5}(\d{2})\d{2}', code)
                    if m_code_cycle:
                        parsed_c = int(m_code_cycle.group(1))
                        if 1 <= parsed_c <= 12:
                            c_num = parsed_c
                            current_detected_cycle = c_num
                    elif code.startswith('INE0') or code.startswith('INO1'):
                        c_num = 1
                    elif code.startswith('INO2'):
                        c_num = 2
                    
                    if c_num not in cycles:
                        cycles[c_num] = []
                    
                    cycles[c_num].append({
                        'esp': int(esp) if esp.isdigit() else 0,
                        'codigo': code or '--',
                        'nombre': name,
                        'asignatura_full': asig_str,
                        'creditos': cred,
                        'tipo': tipo_str,
                        'grupo': grupo_str or '--',
                        'prerequisito': prereq_str or '--'
                    })
                    
    return cycles, metadata

def calculate_course_grade(ep, ec, ef):
    ep = max(0, min(20, int(ep)))
    ec = max(0, min(20, int(ec)))
    ef = max(0, min(20, int(ef)))
    
    exact = 0.30 * ep + 0.40 * ec + 0.30 * ef
    rounded = math.floor(exact + 0.5)
    return exact, rounded

File: js/pdf-parser.js
// pdf-parser.js - Client side PDF parsing using PDF.js

function cleanPdfText(text) {
    if (!text) return "";
    text = text.replace(/\x00/g, '').replace(/\r/g, '');
    text = text.replace(/[\u2010\u2011\u2012\u2013\u2014\u2015\u2212]/g, '-');
    return text;
}

function parseCourseCodeName(asigStr) {
    asigStr = cleanPdfText(asigStr.trim());
    asigStr = asigStr.replace(/^[-\s]+/, '');
    
    let parts = asigStr.split(/\s*-\s*(.+)/);
    if (parts.length >= 2 && /^[A-Za-z0-9]{3,12}$/.test(parts[0].trim())) {
        return [parts[0].trim(), parts[1].trim()];
    }
    
    let m = asigStr.match(/^([A-Za-z0-9]{3,12})\s*-\s*(.*)$/);
    if (m) return [m[1].trim(), m[2].trim()];
    
    let m2 = asigStr.match(/^([A-Za-z0-9]{5,12})\s+(.*)$/);
    if (m2) return [m2[1].trim(), m2[2].trim()];

    return ["", asigStr];
}

function extractHeaderMetadata(textLines) {
    let meta = { student_code: '', student_name: '', facultad: '', escuela: '', plan: '' };
    for (const line of textLines) {
        let line_s = line.trim();
        if ((line_s.includes('Código de Matrícula') || line_s.includes('Cdigo de Matrcula')) && line_s.includes(':')) {
            meta.student_code = line_s.split(':', 2)[1].trim();
        } else if (line_s.includes('Nombres y Apellidos') && line_s.includes(':')) {
            meta.student_name = line_s.split(':', 2)[1].trim();
        } else if (line_s.includes('Facultad') && line_s.includes(':')) {
            meta.facultad = line_s.split(':', 2)[1].trim();
        } else if (line_s.includes('Escuela') && line_s.includes(':')) {
            meta.escuela = line_s.split(':', 2)[1].trim();
        } else if (line_s.includes('Plan') && line_s.includes(':') && (line_s.includes('Plan De') || line_s.includes('202'))) {
            meta.plan = line_s.split(':', 2)[1].trim();
        }
    }
    return meta;
}

function extractCourseCodeFromActa(actaStr) {
    if (!actaStr) return "";
    let m_alpha = actaStr.match(/([A-Za-z]{3,6}\d{3})/);
    if (m_alpha) return m_alpha[1].toUpperCase();
    
    let m_num9 = actaStr.match(/20230(\d{9})/);
    if (m_num9) return m_num9[1];
    
    let m_course9 = actaStr.match(/\b(203\d{6}|\d{9})\b/);
    if (m_course9) {
        let found = m_course9[1];
        if (!found.startsWith('2024') && !found.startsWith('2025') && !found.startsWith('2026')) {
            return found;
        }
    }
    return "";
}

async function getPdfLines(pdfUrlOrData) {
    const loadingTask = pdfjsLib.getDocument(pdfUrlOrData);
    const pdf = await loadingTask.promise;
    
    let allLines = [];
    let isFirstPage = true;
    let metadataLines = [];
    
    for (let i = 1; i <= pdf.numPages; i++) {
        const page = await pdf.getPage(i);
        const textContent = await page.getTextContent();
        
        let itemsByY = {};
        for (const item of textContent.items) {
            let y = Math.round(item.transform[5]);
            if (!itemsByY[y]) itemsByY[y] = [];
            itemsByY[y].push(item);
        }
        
        let yKeys = Object.keys(itemsByY).map(Number).sort((a, b) => b - a);
        let pageLines = [];
        
        for (const y of yKeys) {
            let items = itemsByY[y];
            items.sort((a, b) => a.transform[4] - b.transform[4]);
            
            let lineStr = "";
            let lastX = null;
            for (const it of items) {
                if (lastX !== null && (it.transform[4] - lastX) > 10) {
                    lineStr += " ";
                }
                lineStr += it.str;
                lastX = it.transform[4] + it.width;
            }
            pageLines.push(cleanPdfText(lineStr.trim()));
        }
        
        if (isFirstPage) {
            metadataLines = [...pageLines];
            isFirstPage = false;
        }
        allLines = allLines.concat(pageLines);
    }
    return { lines: allLines, metadataLines };
}

window.parseHistorialFile = async function(file) {
    const arrayBuffer = await file.arrayBuffer();
    const { lines, metadataLines } = await getPdfLines(arrayBuffer);
    
    let periods = [];
    let current_period = null;
    let resumen = {
        required_credits: 221.0, approved_credits: 0.0, obligatorios: 0.0, especialidad: 0.0,
        electivos_generales: 0.0, electivos_especialidad: 0.0, optativos: 0.0, alternativos: 0.0,
        otra_especialidad: 0.0, mas_de_una_vez: 0.0, otros: 0.0, missing_credits: 221.0, ppg: 0.0
    };
    
    let metadata = extractHeaderMetadata(metadataLines);

    for (const l of lines) {
        let l_str = l.trim();
        let nums;
        if (l_str.includes('Creditaje Requerido para Egresar')) {
            nums = l_str.match(/\d+\.?\d*/g); if (nums) resumen.required_credits = parseFloat(nums[nums.length-1]);
        } else if (l_str.includes('Creditaje Apobrado') || l_str.includes('Creditaje Aprobado')) {
            nums = l_str.match(/\d+\.?\d*/g); if (nums) resumen.approved_credits = parseFloat(nums[nums.length-1]);
        } else if (l_str.includes('Obligatorios') && !l_str.includes('Creditaje')) {
            nums = l_str.match(/\d+\.?\d*/g); if (nums) resumen.obligatorios = parseFloat(nums[nums.length-1]);
        } else if (l_str.includes('De Especialidad')) {
            nums = l_str.match(/\d+\.?\d*/g); if (nums) resumen.especialidad = parseFloat(nums[nums.length-1]);
        } else if (l_str.includes('Electivos Generales')) {
            nums = l_str.match(/\d+\.?\d*/g); if (nums) resumen.electivos_generales = parseFloat(nums[nums.length-1]);
        } else if (l_str.includes('Electivos de Especialidad')) {
            nums = l_str.match(/\d+\.?\d*/g); if (nums) resumen.electivos_especialidad = parseFloat(nums[nums.length-1]);
        } else if (l_str.includes('Optativos')) {
            nums = l_str.match(/\d+\.?\d*/g); if (nums) resumen.optativos = parseFloat(nums[nums.length-1]);
        } else if (l_str.includes('Alternativos')) {
            nums = l_str.match(/\d+\.?\d*/g); if (nums) resumen.alternativos = parseFloat(nums[nums.length-1]);
        } else if (l_str.includes('De Otra Especialidad')) {
            nums = l_str.match(/\d+\.?\d*/g); if (nums) resumen.otra_especialidad = parseFloat(nums[nums.length-1]);
        } else if (l_str.includes('Más de una vez') || l_str.includes('Ms de una vez')) {
            nums = l_str.match(/\d+\.?\d*/g); if (nums) resumen.mas_de_una_vez = parseFloat(nums[nums.length-1]);
        } else if (l_str.includes('Otros')) {
            nums = l_str.match(/\d+\.?\d*/g); if (nums) resumen.otros = parseFloat(nums[nums.length-1]);
        } else if (l_str.includes('Creditaje Faltante')) {
            nums = l_str.match(/\d+\.?\d*/g); if (nums) resumen.missing_credits = parseFloat(nums[nums.length-1]);
        } else if (l_str.includes('Promedio Ponderado') && !l_str.includes('General')) {
            nums = l_str.match(/\d+\.\d+/g); if (nums) resumen.ppg = parseFloat(nums[nums.length-1]);
        }
    }

    let pending_course = null;
    let last_added_course = null;
    let unbound_course_header = null;

    for (let line of lines) {
        let line_s = line.replace(/\s+/g, ' ').trim();
        if (!line_s) continue;

        let m_period = line_s.match(/Periodo Acad[eé]mico\s+(\d{4}-\d)/i);
        if (m_period) {
            let p_name = m_period[1];
            current_period = periods.find(p => p.period === p_name);
            if (!current_period) {
                current_period = { period: p_name, courses: [] };
                periods.push(current_period);
            }
            pending_course = null;
            last_added_course = null;
            unbound_course_header = null;
            continue;
        }

        if (!current_period) continue;

        let m_start = line_s.match(/^(\d+)\s+(\d{4})\s+([OEΟ])(?:\s+(.*))?$/);
        if (m_start) {
            let ciclo = m_start[1];
            let plan = m_start[2];
            let tipo = m_start[3];
            let rest = (m_start[4] || '').trim();

            if (unbound_course_header) {
                rest = (unbound_course_header + " " + rest).trim();
                unbound_course_header = null;
            }

            let m_end = rest ? rest.match(/(?:(\d{1,2})\s+)?(\d+\.\d)\s+(\d+)\s+([PA]\s*-\s*\S+)$/) : null;
            if (m_end) {
                let calif = m_end[1];
                let cred = m_end[2];
                let sec = m_end[3];
                let acta = m_end[4];
                let asig = rest.slice(0, m_end.index).trim().replace(/\s+/g, ' ');
                let calif_val = calif ? parseInt(calif) : null;
                
                let [code, name] = parseCourseCodeName(asig);
                if (!code && acta) code = extractCourseCodeFromActa(acta);

                sec = sec || '1';
                let c_id = `${current_period.period}_${code}_${sec}`;
                let existing = current_period.courses.find(c => c.id === c_id);
                if (!existing) {
                    let new_c = {
                        id: c_id, ciclo: parseInt(ciclo), plan: plan, tipo: tipo,
                        codigo: code, nombre: name, asignatura_full: asig,
                        calificacion: calif_val, creditos: parseFloat(cred), seccion: sec, acta: acta,
                        ep: calif_val !== null ? calif_val : 0,
                        ec: calif_val !== null ? calif_val : 0,
                        ef: calif_val !== null ? calif_val : 0,
                        en_curso: calif_val === null, user_edited: false
                    };
                    current_period.courses.push(new_c);
                    last_added_course = new_c;
                } else {
                    last_added_course = existing;
                }
                pending_course = null;
            } else {
                pending_course = { period: current_period, ciclo: parseInt(ciclo), plan: plan, tipo: tipo, prefix: rest };
                last_added_course = null;
            }
        } else if (pending_course) {
            let m_end = line_s.match(/(?:(\d{1,2})\s+)?(\d+\.\d)\s+(\d+)\s+([PA]\s*-\s*\S+)$/);
            if (m_end) {
                let calif = m_end[1];
                let cred = m_end[2];
                let sec = m_end[3];
                let acta = m_end[4];
                
                let suffix = line_s.slice(0, m_end.index).trim();
                let full_asig = (pending_course.prefix + " " + suffix).trim().replace(/\s+/g, ' ');
                let calif_val = calif ? parseInt(calif) : null;
                
                let [code, name] = parseCourseCodeName(full_asig);
                if (!code && acta) code = extractCourseCodeFromActa(acta);

                let target_period = pending_course.period;
                let c_id = `${target_period.period}_${code}_${sec}`;
                let existing = target_period.courses.find(c => c.id === c_id);
                if (!existing) {
                    let new_c = {
                        id: c_id, ciclo: pending_course.ciclo, plan: pending_course.plan, tipo: pending_course.tipo,
                        codigo: code, nombre: name, asignatura_full: full_asig,
                        calificacion: calif_val, creditos: parseFloat(cred), seccion: sec, acta: acta,
                        ep: calif_val !== null ? calif_val : 0,
                        ec: calif_val !== null ? calif_val : 0,
                        ef: calif_val !== null ? calif_val : 0,
                        en_curso: calif_val === null, user_edited: false
                    };
                    target_period.courses.push(new_c);
                    last_added_course = new_c;
                } else {
                    last_added_course = existing;
                }
                pending_course = null;
            } else {
                pending_course.prefix = (pending_course.prefix + " " + line_s).trim();
            }
        } else {
            let has_course_code = /(?:^[A-Za-z0-9]{3,12}\s*-\s*|^\d{9}\s*-\s*)/.test(line_s);
            let is_metadata = ['Página', 'Documento Verificable', 'Ciclo Plan', 'Creditaje', 'Resumen', 'Periodo'].some(k => line_s.includes(k));

            if (has_course_code && !is_metadata) {
                unbound_course_header = line_s;
                last_added_course = null;
            } else if (last_added_course) {
                if (!is_metadata) {
                    let full_asig = (last_added_course.asignatura_full + " " + line_s).trim().replace(/\s+/g, ' ');
                    let [code, name] = parseCourseCodeName(full_asig);
                    let full_code = code || (last_added_course.codigo || '');
                    let full_name = name || full_asig;
                    last_added_course.asignatura_full = full_asig;
                    if (full_code) last_added_course.codigo = full_code;
                    if (full_name) last_added_course.nombre = full_name;
                }
                last_added_course = null;
            }
        }
    }

    return { periods, resumen, metadata };
};

window.parsePlanFile = async function(file) {
    const arrayBuffer = await file.arrayBuffer();
    const loadingTask = pdfjsLib.getDocument(arrayBuffer);
    const pdf = await loadingTask.promise;
    
    let cycles = {};
    let metadata = {};
    let current_detected_cycle = 1;
    let isFirstPage = true;

    for (let i = 1; i <= pdf.numPages; i++) {
        const page = await pdf.getPage(i);
        const textContent = await page.getTextContent();
        
        let itemsByY = {};
        for (const item of textContent.items) {
            let y = Math.round(item.transform[5] / 2) * 2;
            if (!itemsByY[y]) itemsByY[y] = [];
            itemsByY[y].push(item);
        }
        
        let yKeys = Object.keys(itemsByY).map(Number).sort((a, b) => b - a);
        let pageLines = [];
        
        for (const y of yKeys) {
            let items = itemsByY[y];
            items.sort((a, b) => a.transform[4] - b.transform[4]);
            
            let cols = [];
            let currentCol = "";
            let lastX = null;
            
            for (const it of items) {
                if (lastX !== null && (it.transform[4] - lastX) > 15) {
                    cols.push(currentCol.trim());
                    currentCol = "";
                } else if (lastX !== null && (it.transform[4] - lastX) > 3) {
                    currentCol += " ";
                }
                currentCol += it.str;
                lastX = it.transform[4] + it.width;
            }
            if (currentCol) cols.push(currentCol.trim());
            
            let fullStr = cleanPdfText(cols.join("   "));
            pageLines.push(fullStr);
            
            if (cols.length >= 4) {
                let row_str_full = fullStr;
                if (row_str_full.includes('Esp.') && row_str_full.includes('Asignatura')) continue;
                
                let esp = cleanPdfText(cols[0]);
                let asig_str = cleanPdfText(cols[1]).replace(/\n/g, ' ');
                let cred_str = cleanPdfText(cols[2]);
                let tipo_str = cleanPdfText(cols[3]);
                let grupo_str = cols.length > 4 ? cleanPdfText(cols[4]) : '--';
                let prereq_str = cols.length > 5 ? cleanPdfText(cols[5]).replace(/\n/g, ' ').replace(/\s+/g, ' ') : '--';
                
                let cred = parseFloat(cred_str);
                if (!asig_str || !cred_str || isNaN(cred)) continue;
                
                let [code, name] = parseCourseCodeName(asig_str);
                let c_num = current_detected_cycle;
                
                let m_code_cycle = code.match(/\d{5}(\d{2})\d{2}/);
                if (m_code_cycle) {
                    let parsed_c = parseInt(m_code_cycle[1]);
                    if (parsed_c >= 1 && parsed_c <= 12) {
                        c_num = parsed_c;
                        current_detected_cycle = c_num;
                    }
                } else if (code.startsWith('INE0') || code.startsWith('INO1')) {
                    c_num = 1;
                } else if (code.startsWith('INO2')) {
                    c_num = 2;
                }
                
                if (!cycles[c_num]) cycles[c_num] = [];
                
                cycles[c_num].push({
                    esp: !isNaN(parseInt(esp)) ? parseInt(esp) : 0,
                    codigo: code || '--',
                    nombre: name,
                    asignatura_full: asig_str,
                    creditos: cred,
                    tipo: tipo_str,
                    grupo: grupo_str || '--',
                    prerequisito: prereq_str || '--'
                });
            }
        }
        
        if (isFirstPage) {
            metadata = extractHeaderMetadata(pageLines);
            isFirstPage = false;
        }
    }

    return { cycles, metadata };
};
